[FLINK-9885] [FLINK-8101] Finalize Elasticsearch 6.x PR

tzulitai opened the pull request:
## What is the purpose of the change

This PR adds the Elasticsearch 6.x PR, as well as an end-to-end test for it.

## Brief change log

- Cherry-picked @cjolif 's Elasticsearch 5.3+ compatibility and 6.x implementation changes
- Add end to end test for ES 6.x

## Verifying this change

Run the new end-to-end test for ES 6.x.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch6/ElasticsearchSink.java:
```diff
@@ -0,0 +1,91 @@
+ *
+ * @param <T> Type of the elements handled by this sink
+ */
+public class ElasticsearchSink<T> extends ElasticsearchSinkBase<T> {
```
Add `@PublicEvolving`

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch6/ElasticsearchSink.java:
```diff
@@ -0,0 +1,91 @@
+	 *
+	 * @param elasticsearchSinkFunction This is used to generate multiple {@link ActionRequest} from the incoming element
+	 * @param failureHandler This is used to handle failed {@link ActionRequest}
+	 * @param httpHosts The list of {@HttpHost} to which the {@link RestHighLevelClient} connects to.
```
Fix two invalid Javadocs.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch6/ElasticsearchSink.java:
```diff
@@ -0,0 +1,91 @@
+
+import org.apache.http.HttpHost;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.DocWriteRequest;
```
Remove unused import.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/EmbeddedElasticsearchNodeEnvironmentImpl.java:
```diff
@@ -0,0 +1,82 @@
+				.put("http.enabled", false)
+				.put("path.home", tmpDataFolder.getParent())
+				.put("path.data", tmpDataFolder.getAbsolutePath())
+				.put(NetworkModule.HTTP_TYPE_KEY, Netty4Plugin.NETTY_HTTP_TRANSPORT_NAME)
```
Are these values still valid? I thought we are not relying on Netty anymore with the rest client?

tzulitai replied:
You're right, removed.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch6/ElasticsearchSinkITCase.java:
```diff
@@ -0,0 +1,152 @@
+	/**
+	 * Tests that the Elasticsearch sink works properly using a {@link RestHighLevelClient}.
+	 */
+	public void runTransportClientTest() throws Exception {
```
There are no `@Test` annotations to run tests. We should also rename the method in `testXXX` as we usually do it. The super class method names should be updated as `runTransportClientTest` is not correct anymore.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch6/ElasticsearchSinkITCase.java:
```diff
@@ -0,0 +1,152 @@
+		return new ElasticsearchSink<>(userConfig, httpHosts, elasticsearchSinkFunction);
+	}
+
+	private <T> ElasticsearchSinkBase<T> createElasticsearchSink6(
```
Merge this function with `createElasticsearchSink`?

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/test/java/org/apache/flink/streaming/connectors/elasticsearch6/examples/ElasticsearchSinkExample.java:
```diff
@@ -0,0 +1,81 @@
+ * This example shows how to use the Elasticsearch Sink. Before running it you must ensure that
+ * you have a cluster named "elasticsearch" running or change the name of cluster in the config map.
+ */
+public class ElasticsearchSinkExample {
```
We should add tests for our examples.

tzulitai replied:
Actually, we decided to move examples out of the test code. Removing this.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/test/resources/log4j-test.properties:
```diff
@@ -0,0 +1,27 @@
+log4j.appender.testlogger.layout.ConversionPattern=%-4r [%t] %-5p %c %x - %m%n
+
+# suppress the irrelevant (wrong) warnings from the netty channel handler
+log4j.logger.org.jboss.netty.channel.DefaultChannelPipeline=ERROR, testlogger
```
Still necessary?

tzulitai replied:
Removed.

twalthr commented on flink-end-to-end-tests/run-nightly-tests.sh:
```diff
@@ -96,6 +96,7 @@ run_test "Local recovery and sticky scheduling end-to-end test" "$END_TO_END_DIR
 run_test "Elasticsearch (v1.7.1) sink end-to-end test" "$END_TO_END_DIR/test-scripts/test_streaming_elasticsearch.sh 1 https://download.elastic.co/elasticsearch/elasticsearch/elasticsearch-1.7.1.tar.gz"
 run_test "Elasticsearch (v2.3.5) sink end-to-end test" "$END_TO_END_DIR/test-scripts/test_streaming_elasticsearch.sh 2 https://download.elastic.co/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/2.3.5/elasticsearch-2.3.5.tar.gz"
 run_test "Elasticsearch (v5.1.2) sink end-to-end test" "$END_TO_END_DIR/test-scripts/test_streaming_elasticsearch.sh 5 https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-5.1.2.tar.gz"
+run_test "Elasticsearch (v6.3.1) sink end-to-end test" "$END_TO_END_DIR/test-scripts/test_streaming_elasticsearch.sh 6 https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-6.3.1.tar.gz"
```
The test is not runnable on my machine.

```
Elasticsearch node is not running.
grep: /Users/twalthr/flink/flink/build-target/log/*.out: No such file or directory

[FAIL] './test-scripts/test_streaming_elasticsearch.sh 6 https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-6.3.1.tar.gz' failed after 0 minutes and 18 seconds! Test exited with exit code 1
```

The tests exits before elastic search has actually started. Also killing does not work. An Elasticsearch process is still running afterwards.

tzulitai replied:
Added a loop to wait until the Elasticsearch node is really running.

tzulitai commented:
@twalthr I've addressed most of your comments. Thanks a lot for the detailed review.

However, now it seems like the Elasticsearch 6.x IT cases are failing, due to ceased support of Elasticsearch embedded nodes for Java unit tests. I'll have to look a bit more to see how that can be fixed. Will ping you when that is addressed.

cjolif commented on flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch6/Elasticsearch6ApiCallBridge.java:
```diff
@@ -0,0 +1,109 @@
+
+	@Override
+	public RestHighLevelClient createClient(Map<String, String> clientConfig) {
+		RestHighLevelClient rhlClient =
```
might have been good to support:
- context path / path prefix in addition to host
- login/password for Elasticsearch instances protected

that's ok to not do it as soon as the user can by subclassing. Maybe to make it easier to subclass there should be two methods.

keep the public `createClient` one that returns the `RHLClient`. And add a protected method `createRestClientBuilder` which return the `RestClientBuilder`. This way one can just redefine the protected method and let the pubic one handle the actual `RHLClient` instanciation from the `RestClientBuilder` created by the protected method.

tzulitai replied:
Thanks @cjolif. Your suggestion to allow overrides sound good, I'll add that.
Either way, I'll still try to add in the features you mentioned (IIRC, you already provided some code snippets for that somewhere).

tzulitai commented:
Hi @cjolif, @twalthr, this PR is ready for a second pass now.

- End-to-end tests now pass stably.
- Addresses all of @twalthr's comments
- As for @cjolif's comment on custom configuration for the REST client, I addressed this by introducing a `RestClientFactory` that offers opportunity for the user to tweak the `RestClientBuilder`
- I also decided to use a Builder pattern for instantiating the `ElasticsearchSink`. Along with the new `RestClientFactory`, we would have yet more constructors, so I felt that using a builder pattern would be a better idea for the future.

tzulitai commented on docs/dev/connectors/elasticsearch.md:
```diff
@@ -151,22 +151,47 @@ List<HttpHost> httpHost = new ArrayList<>();
-        indexer.add(createIndexRequest(element));
+Map<String, String> bulkRequestsConfig = new HashMap<>();
+// This instructs the sink to emit after every element, otherwise they would be buffered
+bulkRequestsConfig.put("bulk.flush.max.actions", "1");
```
Correct.

twalthr commented on flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkTestBase.java:
```diff
@@ -189,4 +191,10 @@ public void runInvalidElasticsearchClusterTest() throws Exception {
+	 * Creates a version-specific Elasticsearch sink to connect to a specific Elasticsearch node.
+	 */
+	protected abstract ElasticsearchSinkBase<Tuple2<Integer, String>, C> createElasticsearchSinkForNode(
+		Map<String, String> userConfig, ElasticsearchSinkFunction<Tuple2<Integer, String>> elasticsearchSinkFunction, String ipAddress) throws Exception;
```
One line per parameter? Also the line above.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch6/Elasticsearch6ApiCallBridge.java:
```diff
@@ -65,12 +66,16 @@
 
 		RestHighLevelClient rhlClient = new RestHighLevelClient(builder);
 
+		if (!rhlClient.ping()) {
```
Should we add some logging here for information that we perform a communication step?

tzulitai replied:
Good idea.

twalthr commented on docs/dev/connectors/elasticsearch.md:
```diff
@@ -198,10 +198,6 @@ List<HttpHost> httpHost = new ArrayList<>();
 httpHosts.add(new HttpHost("127.0.0.1", 9200, "http"));
 httpHosts.add(new HttpHost("10.2.3.1", 9200, "http"));
 
-Map<String, String> bulkRequestsConfig = new HashMap<>();
```
Remove `Map` from the imports.

tzulitai replied:
It is still used when processing requests.
It can be removed from the Scala code snippet, though.

twalthr commented on flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch6/ElasticsearchSink.java:
```diff
@@ -97,6 +97,15 @@ public Builder(List<HttpHost> httpHosts, ElasticsearchSinkFunction<T> elasticsea
+		 * @param bulkRequestsConfig user configuration to configure bulk flushing behaviour.
+		 */
+		public void setBulkRequestsConfig(Map<String, String> bulkRequestsConfig) {
+			this.bulkRequestsConfig = bulkRequestsConfig;
```
Check for null. But do we actually need to expose the raw map if it is not used in ES 6 and later?

tzulitai replied:
My thought was just that it's more or less a convenience for users. In some scenarios, having a properties like setter would be easier to use. I'm completely ok with removing it, though.

tzulitai commented:
@twalthr I've addressed your latest comments. Could you have another pass?

tzulitai commented:
Thanks @twalthr, I've opened an issue to tackle the ITCases for 5.x and 6.x here:
https://issues.apache.org/jira/browse/FLINK-10015

tzulitai commented:
Travis gives a green light:
https://travis-ci.org/tzulitai/flink/builds/410657354

will merge this

joyang1 commented:
@tzulitai  i use flink-connector-elasticsearch6_2.11 have exception: org.apache.flink.runtime.client.JobExecutionException: TimerException{java.lang.NoSuchMethodError: org.elasticsearch.action.bulk.BulkProcessor.add(Lorg/elasticsearch/action/ActionRequest;)Lorg/elasticsearch/action/bulk/BulkProcessor;}

how I can  solve this?

twalthr commented:
@joyang1 please post this on the user ML or open a jira for it if you think this is a bug.
